---
title: "RNAseq_analyse"
author: "mason-YHY"
date: "8/30/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

载入数据
------------------------------------------------------------
```{r, load data}
library(edgeR)
x <- read.table("expr.txt", header=T, sep="\t", row.name=1)
```

过滤低reads数基因,并生成分组信息
```{r, filter low reads and make group}
df <- x[rowSums(x)>0, ]
dg <- c(rep("control",2),rep("case",2))
dl <- DGEList(counts=df, group=dg)
dn <- calcNormFactors(dl, method="TMM")
```


分析差异基因
------------------------------------------------------------
```{r}
dc <- estimateCommonDisp(dn)
de <- exactTest(dc,pair=c("control", "case"))
top <- topTags(de,n=length(de$table$PValue))
topFDR <- subset(top$table,FDR < 0.05 & abs(logFC) >= 1)
```

调整差异分析结果格式
```{r}
topFDR$trend <- "UP"
topFDR[topFDR$logFC < 0,]$trend <- "DOWN"
detags <- rownames(topFDR)
topFDR <- cbind(df[detags,],topFDR)
topFDR <- data.frame(GID=row.names(topFDR),topFDR)
```

可视化差异分析
---------------------------------------------
```{r}
detags <- rownames(topFDR)

#log2-counts-per-million (logCPM)
plotMD(de, de.tags=detags, main="FC plot")
abline(h=c(-1,1), col="dodgerblue")
```

绘制火山图
```{r}
data <-top$table
color=c()
for (i in 1:nrow(data)){
  if (data$FDR[i] < 0.05 & data$logFC[i] >1){
    col="red"
  } else if (data$FDR[i] < 0.05 & data$logFC[i]<(-1)){
    col="green"
  } else {col="black"}
  color=c(color,col)
}
plot(data$logFC,-log10(data$FDR),col=color,xlab="logFC",ylab="-log10(FDR)")
abline(h=-log10(0.05),v=c(-1,1),lty=2)
```

绘制热图
```{r}
library(pheatmap)
data <- cpm(df)
data <- data[match(row.names(topFDR),row.names(data)),]
mat <- as.matrix(data)
pheatmap(log(mat+1), color = colorRampPalette(c("#1409a4", "#81ebe5", "#3df134", "#f4f70c","#fb080e"))(50), border_color = "NA", show_rownames=F)
```

GO富集分析
--------------------------------------------
```{r}
library(clusterProfiler)
library(org.Hs.eg.db)
diff.gene <- topFDR$GID
go.BP <-  enrichGO(diff.gene, 
                   OrgDb = "org.Hs.eg.db", 
                   ont='BP',
                   pvalueCutoff = 0.05, 
                   qvalueCutoff = 0.2, 
                   keyType = 'ENTREZID')
head(go.BP)
```


```{r}
data(geneList, package = "DOSE")
de <- names(geneList)[1:100]
yy <- enrichGO(de, 'org.Hs.eg.db', ont="BP", pvalueCutoff=0.01)
head(yy)
```

可视化GO富集分析结果
```{r}
dotplot(yy)
```

KEGG富集分析
```{r}
kegg <- enrichKEGG(de, organism = 'hsa', keyType = 'kegg', pvalueCutoff = 0.05, use_internal_data = FALSE)
head(kegg)
```
可视化通路
```{r, message=F}
#browseKEGG(kegg, "hsa00230")
#library("pathview")
#hsa04110 <- pathview(gene.data  = geneList,
#                     pathway.id = "hsa00230",
#                     species    = "hsa",
#                     limit      = list(gene=max(abs(geneList)), cpd=1))
```

