---
title: "report"
author: "YHY-mason"
date: "6/20/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

load data(Enterotypes of the human gut microbiome (2011))
--------------------
```{r example_bio}
library(phyloseq)
library(ggplot2)
theme_set(theme_bw())

rich_dense_biom  = system.file("extdata", "rich_dense_otu_table.biom",  package="phyloseq")
rich_sparse_biom = system.file("extdata", "rich_sparse_otu_table.biom", package="phyloseq")
min_dense_biom   = system.file("extdata", "min_dense_otu_table.biom",   package="phyloseq")
min_sparse_biom  = system.file("extdata", "min_sparse_otu_table.biom",  package="phyloseq")
treefilename = system.file("extdata", "biom-tree.phy",  package="phyloseq")
refseqfilename = system.file("extdata", "biom-refseq.fasta",  package="phyloseq")

exampl_biom <- import_biom(rich_dense_biom, treefilename, refseqfilename, parseFunction=parse_taxonomy_greengenes)

str(exampl_biom)
```

```{r plot}
plot_bar(exampl_biom, fill = "Family")
plot_bar(exampl_biom, fill="Genus")

plot_tree(exampl_biom, color="Genus", shape="BODY_SITE", size="abundance", label.tips="taxa_names")
plot_heatmap(exampl_biom)

plot_richness(exampl_biom, x="BODY_SITE", color="Description")
```

```{r GlobalPatterns}
data(GlobalPatterns)
str(GlobalPatterns)

#plot_bar(GlobalPatterns, fill = "Family")
```

Remove OTUs 
--------------------------
```{r}
#
GP = GlobalPatterns
wh0 = genefilter_sample(GP, filterfun_sample(function(x) x > 10), A=0.5*nsamples(GP))
GP1 = prune_taxa(wh0, GP)
#
#GP1 = transform_sample_counts(GP1, function(x) 1E6 * x/sum(x))
#增加是否为human的新样本分组
human = get_variable(GP1, "SampleType") %in% c("Feces", "Mock", "Skin", "Tongue")
sample_data(GP1)$human <- factor(human)
```

计算筛选otu top5
-------------------
```{r}
phylum.sum = tapply(taxa_sums(GP1), tax_table(GP1)[, "Phylum"], sum, na.rm=TRUE)
top5phyla = names(sort(phylum.sum, TRUE))[1:5]
GP1 = prune_taxa((tax_table(GP1)[, "Phylum"] %in% top5phyla), GP1)
```


```{r}
GP.ord <- ordinate(GP1, "NMDS", "bray")
p1 = plot_ordination(GP1, GP.ord, type="taxa", color="Phylum", title="taxa")
print(p1)

p1 + facet_wrap(~Phylum)

p2 = plot_ordination(GP1, GP.ord, type="samples", color="SampleType", shape="human")
print(p2)
p2 + geom_polygon(aes(fill=SampleType)) + geom_point(size=5) + ggtitle("samples")

p3 = plot_ordination(GP1, GP.ord, type="biplot", color="SampleType", shape="Phylum", title="biplot")
print(p3)
```


```{r}
plot_richness(GP1)

plot_richness(GP1, x="human", color="SampleType", measures=c("Chao1", "Shannon"))
```

```{r}
GP1 <- prune_taxa(taxa_names(GlobalPatterns)[1:50], GlobalPatterns)
plot_tree(GP1)

p4 <- plot_tree(GP1, nodelabf=nodeplotboot(), ladderize="left", color="SampleType",label.tips="taxa_names")

print(p4)

p4+coord_polar(theta="y")
```



