---
title: "methylation"
author: "mason-YHY"
date: "8/29/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

安装并载入分析所需R包
---------------------
```{r libload, results="hide", echo=FALSE, message=FALSE, warning=FALSE}
library(limma)
library(minfi)
library(DMRcate)

library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(IlluminaHumanMethylation450kmanifest)
library(RColorBrewer)
library(missMethyl)
library(matrixStats)
library(minfiData)
```

读入数据[GSE49667](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE49667)
-------------------------------------------
```{r}
mVals <- read.csv("mvalue.csv", row.names = 1)
mVals <- as.matrix(mVals)
targets <- read.csv("sample_sheet.csv")

ann450k <-  getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)

```


甲基化位点差异分析
--------------------------------------------
基于5％的错误发现率，在“native”与“rTreg”比较中存在3021个显着差异甲基化的CpG，而“rTreg”与“act_rTreg”没有显示任何显着的差异甲基化。
```{r dmps}
# this is the factor of interest
cellType <- factor(targets$Sample_Group)
# this is the individual effect that we need to account for
individual <- factor(targets$Sample_Source) 

# use the above to create a design matrix
design <- model.matrix(~0+cellType+individual, data=targets)
colnames(design) <- c(levels(cellType),levels(individual)[-1])
 
# fit the linear model 
fit <- lmFit(mVals, design)
# create a contrast matrix for specific comparisons
contMatrix <- makeContrasts(naive-rTreg,
                           naive-act_naive,
                           rTreg-act_rTreg,
                           act_naive-act_rTreg,
                           levels=design)
contMatrix

# fit the contrasts
fit2 <- contrasts.fit(fit, contMatrix)
fit2 <- eBayes(fit2)

# look at the numbers of DM CpGs at FDR < 0.05
summary(decideTests(fit2))
```

```{r annotatedmps}
# get the table of results for the first contrast (naive - rTreg)
ann450kSub <- ann450k[match(rownames(mVals),ann450k$Name),
                      c(1:4,12:19,24:ncol(ann450k))]
DMPs <- topTable(fit2, num=Inf, coef=1, genelist=ann450kSub)
head(DMPs)
```

DMR（Differential Methylation Regions）甲基化区域差异化分析&可视化
--------------------------------------------------------
- charm中的dmrFind
- minfi中的bumphunter
- DMRcate中的dmrcate

它们各自基于不同的统计方法。根据经验，“bumphunter”和“dmrFind”函数运行起来有点慢，除非你有计算机基础设施来并行化，因为它们使用排列来分配重要性。在此工作流程中，我们将使用`dmrcate`执行分析。因为它基于* limma *，我们可以直接使用我们之前定义的`design`和`contMatrix`。

先用DMRcate包的cpg.annotate函数做单个甲基化位点的差异分析以及注释

```{r cpgannotate}
myAnnotation <- cpg.annotate(object = mVals, datatype = "array", what = "M", 
                             analysis.type = "differential", design = design, 
                             contrasts = TRUE, cont.matrix = contMatrix, 
                             coef = "naive - rTreg", arraytype = "450K")
str(myAnnotation)
```

接着用dmrcate函数在上面单个甲基化位点分析的基础上进行合并，从而识别出DMR
```{r dmrcate, message=FALSE}
#endif /* NEWSTUFF */
DMRs <- dmrcate(myAnnotation, lambda=1000, C=2)
head(DMRs$results)
```

用DMR.plot函数进行可视化，看看这些DMRs在基因组上的展现形式
```{r resultsranges}
# convert the regions to annotated genomic ranges
data(dmrcatedata)
results.ranges <- extractRanges(DMRs, genome = "hg19")

# set up the grouping variables and colours
pal <- brewer.pal(8,"Dark2")
groups <- pal[1:length(unique(targets$Sample_Group))]
names(groups) <- levels(factor(targets$Sample_Group))
cols <- groups[as.character(factor(targets$Sample_Group))]
samps <- 1:nrow(targets)
```

该图显示了DMRcate分析确定的排名最高的DMR
```{r figure10, fig.width=10, fig.height=10, fig.cap="The DMRcate \"DMR.plot\" function allows you to quickly visualise DMRs in their genomic context. By default, the plot shows the location of the DMR in the genome, the position of any genes that are nearby, the base pair positions of the CpG probes, the methylation levels of the individual samples as a heatmap and the mean methylation levels for the various sample groups in the experiment. This plot shows the top ranked DMR identified by the DMRcate analysis."}
# draw the plot for the top DMR
par(mfrow=c(1,1))
DMR.plot(ranges=results.ranges, dmr=1, CpGs=mVals, phen.col=cols, what = "M",
         arraytype = "450K", pch=16, toscale=TRUE, plotmedians=TRUE, 
         genome="hg19", samps=samps)
```